File: benchmarks/weird_captcha_gym/shared_runtime/app/mechanics/exit_vim_terminal_escape.js
(() => {
  "use strict";

  let helpersCache = null;
  let model = null;
  let keyHandler = null;

  function esc(value) {
    return String(value == null ? "" : value)
      .replaceAll("&", "&amp;")
      .replaceAll("<", "&lt;")
      .replaceAll(">", "&gt;")
      .replaceAll('"', "&quot;");
  }

  function layerName() {
    if (model.layerIndex < 0) return "editor";
    if (model.layerIndex >= model.layerOrder.length) return "complete";
    return model.layerOrder[model.layerIndex];
  }

  function modeName() {
    return model.layerIndex < 0 ? model.mode : layerName();
  }

  function clampCursor() {
    model.row = Math.max(0, Math.min(model.row, model.buffer.length - 1));
    model.col = Math.max(0, Math.min(model.col, model.buffer[model.row].length));
  }

  function pushUndo() {
    model.undo.push({buffer: [...model.buffer], row: model.row, col: model.col});
    model.undo = model.undo.slice(-30);
  }

  function advanceLayer() {
    const current = layerName();
    if (current === "editor" || current === "complete") return;
    model.exitLog.push(current);
    model.layerIndex += 1;
    model.sshInput = "";
    model.message = model.layerIndex >= model.layerOrder.length ? "ALL SHELLS CLOSED" : `RETURNED TO ${layerName().toUpperCase()}`;
  }

  function applyEditorKey(key, ctrl, alt, meta) {
    if (model.mode === "insert") {
      if (key === "Escape") {
        model.mode = "normal";
        model.pending = "";
        model.message = "NORMAL MODE";
      } else if (key === "Backspace") {
        if (model.col > 0) {
          pushUndo();
          const line = model.buffer[model.row];
          model.buffer[model.row] = line.slice(0, model.col - 1) + line.slice(model.col);
          model.col -= 1;
        }
      } else if (key === "Delete") {
        const line = model.buffer[model.row];
        if (model.col < line.length) {
          pushUndo();
          model.buffer[model.row] = line.slice(0, model.col) + line.slice(model.col + 1);
        }
      } else if (key.length === 1 && !ctrl && !alt && !meta) {
        pushUndo();
        const line = model.buffer[model.row];
        model.buffer[model.row] = line.slice(0, model.col) + key + line.slice(model.col);
        model.col += 1;
        model.insertedChars += 1;
      }
      clampCursor();
      return;
    }

    if (model.mode === "command") {
      if (key === "Escape") {
        model.mode = "normal";
        model.command = "";
        model.message = "COMMAND CANCELLED";
      } else if (key === "Backspace") {
        model.command = model.command.slice(0, -1);
      } else if (key === "Enter") {
        const command = model.command;
        model.commandHistory.push(command);
        model.command = "";
        model.mode = "normal";
        if (command === "wq" && model.buffer.every((line, index) => line === model.targetBuffer[index])) {
          model.saved = true;
          model.layerIndex = 0;
          model.message = `MANIFEST WRITTEN · ENTERING ${layerName().toUpperCase()}`;
        } else {
          model.commandErrors += 1;
          model.message = command === "wq" ? "E492: MANIFEST STILL CORRUPT" : `E492: NOT AN EDITOR COMMAND: ${command || "<empty>"}`;
        }
      } else if (key.length === 1 && !ctrl && !alt && !meta) {
        model.command += key;
      }
      return;
    }

    if (key === "Escape") {
      model.pending = "";
      model.message = "NORMAL MODE";
      return;
    }
    if (ctrl || alt || meta) {
      model.pending = "";
      return;
    }
    if (key === ":") {
      model.mode = "command";
      model.command = "";
      model.pending = "";
      model.message = "COMMAND LINE";
      return;
    }
    if (key === "g") {
      if (model.pending === "g") {
        model.row = 0;
        model.col = 0;
        model.pending = "";
        model.message = "TOP OF BUFFER";
      } else {
        model.pending = "g";
        model.message = "g_";
      }
      return;
    }
    if (key === "c") {
      if (model.pending === "c") {
        pushUndo();
        model.buffer[model.row] = "";
        model.col = 0;
        model.mode = "insert";
        model.pending = "";
        model.clearCount += 1;
        model.message = "-- INSERT -- / LINE CLEARED";
      } else {
        model.pending = "c";
        model.message = "c_ / AWAIT MOTION";
      }
      return;
    }
    model.pending = "";
    if (key === "j" || key === "ArrowDown") model.row += 1;
    else if (key === "k" || key === "ArrowUp") model.row -= 1;
    else if (key === "h" || key === "ArrowLeft") model.col -= 1;
    else if (key === "l" || key === "ArrowRight") model.col += 1;
    else if (key === "0" || key === "Home") model.col = 0;
    else if (key === "$" || key === "End") model.col = model.buffer[model.row].length;
    else if (key === "i") {
      model.mode = "insert";
      model.message = "-- INSERT --";
    } else if (key === "a") {
      model.col = Math.min(model.buffer[model.row].length, model.col + 1);
      model.mode = "insert";
      model.message = "-- INSERT --";
    } else if (key === "x") {
      const line = model.buffer[model.row];
      if (model.col < line.length) {
        pushUndo();
        model.buffer[model.row] = line.slice(0, model.col) + line.slice(model.col + 1);
      }
    } else if (key === "u" && model.undo.length) {
      const snapshot = model.undo.pop();
      model.buffer = [...snapshot.buffer];
      model.row = snapshot.row;
      model.col = snapshot.col;
      model.message = "1 CHANGE UNDONE";
    }
    clampCursor();
  }

  function applyKey(key, ctrl, alt, meta) {
    const current = layerName();
    if (current === "editor") applyEditorKey(key, ctrl, alt, meta);
    else if (current === "pager") {
      if (key.toLowerCase() === "q" && !ctrl && !alt && !meta) advanceLayer();
      else model.message = "PAGER CAPTURES INPUT · q QUITS";
    } else if (current === "job") {
      if (key.toLowerCase() === "c" && ctrl && !alt && !meta) advanceLayer();
      else model.message = "FOREGROUND JOB STILL RUNNING · ^C INTERRUPTS";
    } else if (current === "ssh") {
      if (key === "Backspace") model.sshInput = model.sshInput.slice(0, -1);
      else if (key === "Enter") {
        if (model.sshInput.trim() === "exit") advanceLayer();
        else {
          model.commandErrors += 1;
          model.message = `${model.sshInput || "<empty>"}: command not found`;
          model.sshInput = "";
        }
      } else if (key.length === 1 && !ctrl && !alt && !meta) model.sshInput += key;
    }
  }

  function recordKey(event) {
    if (!model || model.submitting || model.terminal || event.repeat) return;
    if (event.target?.closest?.(".cheat-panel")) return;
    const key = String(event.key || "");
    if (!key) return;
    event.preventDefault();
    clearFreshFailure();
    const item = {
      sequence: model.events.length + 1,
      key,
      ctrl: Boolean(event.ctrlKey),
      shift: Boolean(event.shiftKey),
      alt: Boolean(event.altKey),
      meta: Boolean(event.metaKey),
      layer_before: layerName(),
      mode_before: modeName(),
      layer_after: "",
      mode_after: "",
    };
    applyKey(item.key, item.ctrl, item.alt, item.meta);
    item.layer_after = layerName();
    item.mode_after = modeName();
    model.events.push(item);
    renderTerminal();
    if (layerName() === "complete") window.setTimeout(() => submit(), 140);
  }

  function clearFreshFailure() {
    const root = document.querySelector(".terminal-escape");
    if (!root || root.dataset.freshFailure !== "true") return;
    root.dataset.freshFailure = "false";
    document.querySelector(".terminal-fail-stamp")?.remove();
    helpersCache.setReadout("SESSION ACTIVE · KEYSTROKES RECORDED", "idle");
  }

  function cursorLine(line, row) {
    if (row !== model.row || model.layerIndex >= 0) return esc(line);
    const col = Math.max(0, Math.min(model.col, line.length));
    const before = esc(line.slice(0, col));
    const current = esc(line[col] || " ");
    const after = esc(line.slice(col + (col < line.length ? 1 : 0)));
    return `${before}<span class="terminal-cursor">${current}</span>${after}`;
  }

  function layerStackMarkup() {
    const layers = ["editor", ...model.layerOrder];
    return layers.map((layer, index) => {
      let status = "queued";
      if (model.layerIndex < 0) status = index === 0 ? "active" : "queued";
      else if (index === 0 || index - 1 < model.layerIndex) status = "done";
      else if (index - 1 === model.layerIndex) status = "active";
      return `<li data-status="${status}"><i>${String(index + 1).padStart(2, "0")}</i><b>${esc(layer.toUpperCase())}</b><span>${status.toUpperCase()}</span></li>`;
    }).join("");
  }

  function editorMarkup() {
    const lines = model.buffer.map((line, index) => `<div class="terminal-buffer-line${index === model.row ? " is-current" : ""}"><i>${String(index + 1).padStart(2, "0")}</i><code>${cursorLine(line, index)}</code><span>${line === model.targetBuffer[index] ? "CLEAN" : "CORRUPT"}</span></div>`).join("");
    const command = model.mode === "command" ? `:${esc(model.command)}<span class="terminal-cursor"> </span>` : esc(model.message);
    return `<div class="terminal-editor-bar"><span>manifest.cfg</span><b>${model.saved ? "WRITTEN" : "[+] UNSAVED"}</b></div><div class="terminal-buffer">${lines}</div><div class="terminal-command"><b>${esc(model.mode.toUpperCase())}</b><code>${command}</code><span>${model.row + 1}:${model.col + 1}</span></div>`;
  }

  function pagerMarkup() {
    return `<div class="terminal-layer-screen pager-screen"><header>VERIFICATION HANDBOOK(1)</header><p>SESSION WRAPPER NOTICE</p><p>The corrected manifest has returned into a read-only pager.</p><p>Closing the browser does not close the pager.</p><p>Scroll state is irrelevant; terminate this modal layer explicitly.</p><p class="dim">checksum ${esc(model.state.challenge_id)} · page 1/1</p><footer><b>--More-- (END)</b><span>q quits</span></footer></div>`;
  }

  function jobMarkup() {
    return `<div class="terminal-layer-screen job-screen"><header>foreground: audit-stream --follow</header><ol><li>[ok] manifest write observed</li><li>[ok] editor process exited</li><li>[wait] streaming verifier heartbeat…</li><li>[wait] foreground job owns the terminal</li></ol><div class="job-pulse"><i></i><i></i><i></i><i></i><i></i></div><footer><b>RUNNING</b><span>interrupt with ^C</span></footer></div>`;
  }

  function sshMarkup() {
    return `<div class="terminal-layer-screen ssh-screen"><p>Connection to ${esc(model.state.host)} is still open.</p><p>Last login: seed-bound console / pseudo-terminal 1</p><div class="ssh-prompt"><b>verify@${esc(model.state.host)}:~$</b><code>${esc(model.sshInput)}<span class="terminal-cursor"> </span></code></div><footer>type <b>exit</b>, then press Enter</footer></div>`;
  }

  function renderTerminal() {
    if (!model) return;
    const viewport = document.getElementById("terminal-viewport");
    const current = layerName();
    if (viewport) {
      if (current === "editor") viewport.innerHTML = editorMarkup();
      else if (current === "pager") viewport.innerHTML = pagerMarkup();
      else if (current === "job") viewport.innerHTML = jobMarkup();
      else if (current === "ssh") viewport.innerHTML = sshMarkup();
      else viewport.innerHTML = '<div class="terminal-layer-screen complete-screen"><span>TTY RELEASED</span><strong>VERIFYING…</strong></div>';
      viewport.dataset.layer = current;
    }
    const stack = document.getElementById("terminal-layer-stack");
    if (stack) stack.innerHTML = layerStackMarkup();
    const mode = document.getElementById("terminal-mode-state");
    if (mode) mode.textContent = modeName().toUpperCase();
    const eventCount = document.getElementById("terminal-event-count");
    if (eventCount) eventCount.textContent = String(model.events.length).padStart(3, "0");
    const depth = document.getElementById("terminal-depth-count");
    if (depth) depth.textContent = `${Math.min(model.exitLog.length, model.layerOrder.length)} / ${model.layerOrder.length}`;
    if (!model.submitting && !model.terminal) {
      const label = current === "editor" ? `${model.mode.toUpperCase()} MODE · BUFFER ${model.saved ? "WRITTEN" : "DIRTY"}` : current === "complete" ? "ALL LAYERS CLOSED" : `${current.toUpperCase()} LAYER ACTIVE`;
      helpersCache.setReadout(label, "idle");
    }
  }

  function finalState() {
    return {
      buffer: [...model.buffer],
      saved: model.saved,
      layer_index: model.layerIndex,
      exit_log: [...model.exitLog],
      mode: modeName(),
    };
  }

  async function submit() {
    if (!model || model.submitting || model.terminal) return;
    model.submitting = true;
    document.getElementById("terminal-verify")?.setAttribute("disabled", "disabled");
    helpersCache.setReadout("REPLAYING MODES / BUFFER / SHELL EXITS…", "pending");
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify({
          mechanic_id: model.state.mechanic_id,
          challenge_id: model.state.challenge_id,
          events: model.events,
          final_state: finalState(),
          completed: true,
        }),
      });
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.terminal = true;
        const root = document.querySelector(".terminal-escape");
        root?.classList.add("is-pass");
        root?.insertAdjacentHTML("beforeend", '<div class="terminal-pass-stamp"><span>ALL MODAL LAYERS RELEASED</span><strong>PASS</strong><i>KEY REPLAY VERIFIED</i></div>');
        helpersCache.setReadout("PASS", "passed");
      } else if (outcome.passed === false && outcome.state) {
        await helpersCache.render(outcome.state);
        const root = document.querySelector(".terminal-escape");
        if (root) root.dataset.freshFailure = "true";
        root?.insertAdjacentHTML("beforeend", '<div class="terminal-fail-stamp"><span>SESSION ESCAPE REJECTED</span><strong>FAIL</strong><i>FRESH TTY ISSUED</i></div>');
        helpersCache.setReadout("FAIL · FRESH TTY ISSUED", "error");
      } else {
        model.submitting = false;
        document.getElementById("terminal-verify")?.removeAttribute("disabled");
        helpersCache.setReadout("FAIL · REPLAY UNAVAILABLE", "error");
      }
    } catch (_error) {
      model.submitting = false;
      document.getElementById("terminal-verify")?.removeAttribute("disabled");
      helpersCache.setReadout("FAIL · TERMINAL LINK OFFLINE", "error");
    }
  }

  async function render(state, helpers) {
    helpersCache = helpers || helpersCache;
    if (!helpersCache) throw new Error("exit_vim_terminal_escape requires runtime helpers");
    if (keyHandler) window.removeEventListener("keydown", keyHandler, true);
    document.body.dataset.mechanic = "modal-terminal-escape";
    document.body.dataset.cheatMode = helpersCache.isCheatMode() ? "true" : "false";
    model = {
      state,
      buffer: [...(state.initial_buffer || [])],
      targetBuffer: [...(state.target_buffer || [])],
      layerOrder: [...(state.layer_order || [])],
      layerIndex: -1,
      mode: "normal",
      row: 0,
      col: 0,
      pending: "",
      command: "",
      undo: [],
      saved: false,
      exitLog: [],
      sshInput: "",
      clearCount: 0,
      insertedChars: 0,
      commandErrors: 0,
      commandHistory: [],
      message: "NORMAL MODE",
      events: [],
      submitting: false,
      terminal: false,
    };
    window.exitVimTerminalModel = model;
    helpersCache.app.innerHTML = `<section class="terminal-escape" data-challenge-id="${esc(state.challenge_id)}" data-fresh-failure="false" tabindex="0">
      <header class="terminal-head"><div><span>MODAL TERMINAL ESCAPE / ${esc(state.session_label)}</span><h1>${esc(state.prompt)}</h1></div><aside><span>ACTIVE MODE</span><b id="terminal-mode-state">NORMAL</b><i>KEY EVENTS <strong id="terminal-event-count">000</strong></i></aside></header>
      <main class="terminal-main">
        <section class="terminal-frame"><div class="terminal-chrome"><i></i><i></i><i></i><span>root@${esc(state.host)} — 96×28</span></div><div class="terminal-viewport" id="terminal-viewport" data-layer="editor"></div><div class="terminal-focus-note">CLICK TERMINAL TO FOCUS · PASTE DISABLED</div></section>
        <aside class="terminal-brief">
          <div class="terminal-brief-title"><span>RECOVERY MANIFEST</span><i>POSTED TARGET</i></div>
          <ol class="terminal-targets">${(state.target_buffer || []).map((line, index) => `<li><i>${String(index + 1).padStart(2, "0")}</i><code>${esc(line)}</code></li>`).join("")}</ol>
          <div class="terminal-keystroke-card"><span>EDITOR FIELD PROCEDURE</span><p><kbd>gg</kbd> top · <kbd>j/k</kbd> line</p><p><kbd>cc</kbd> replace line · type target</p><p><kbd>Esc</kbd> normal · <kbd>:wq ↵</kbd> write/quit</p><small><kbd>u</kbd> undoes the latest edit</small></div>
          <div class="terminal-layer-title"><span>NESTED MODAL STACK</span><b>ESCAPED <i id="terminal-depth-count">0 / ${state.layer_order.length}</i></b></div>
          <ol class="terminal-layer-stack" id="terminal-layer-stack"></ol>
          <div class="terminal-outer-help"><span>OUTER ESCAPES APPEAR WHEN ACTIVE</span><p>PAGER <kbd>q</kbd> · JOB <kbd>Ctrl+C</kbd> · SSH <kbd>exit ↵</kbd></p></div>
        </aside>
      </main>
      <footer class="terminal-foot"><div class="readout" data-status="idle">NORMAL MODE · BUFFER DIRTY</div><span>RAW KEYDOWN REPLAY / NO FINAL-STATE TRUST</span><button id="terminal-verify" type="button">${esc(state.submit_label || "VERIFY SESSION")} →</button></footer>
      ${helpersCache.cheatPanelTemplate()}
    </section>`;
    const root = document.querySelector(".terminal-escape");
    root?.addEventListener("click", (event) => {
      if (!event.target.closest("button, input")) root.focus();
    });
    root?.addEventListener("paste", (event) => event.preventDefault());
    root?.addEventListener("drop", (event) => event.preventDefault());
    document.getElementById("terminal-verify")?.addEventListener("click", submit);
    keyHandler = recordKey;
    window.addEventListener("keydown", keyHandler, true);
    renderTerminal();
    root?.focus();
    helpersCache.installCheatPanel();
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.exit_vim_terminal_escape = {rootSelector: ".terminal-escape", render};
})();

File: benchmarks/weird_captcha_gym/shared_runtime/server/incubator_graders/exit_vim_terminal_escape.py
from __future__ import annotations

from copy import deepcopy
from typing import Any


MECHANIC_ID = "exit_vim_terminal_escape"


def _layer_name(state: dict[str, Any]) -> str:
    index = int(state["layer_index"])
    if index < 0:
        return "editor"
    layers = state["layer_order"]
    if index >= len(layers):
        return "complete"
    return str(layers[index])


def _mode_name(state: dict[str, Any]) -> str:
    return str(state["mode"]) if int(state["layer_index"]) < 0 else _layer_name(state)


def _editor_lines(state: dict[str, Any]) -> list[str]:
    index = int(state["buffer_index"])
    if index == 0:
        return state["buffer"]
    return state["reference_buffers"][index - 1]["lines"]


def _editor_writable(state: dict[str, Any]) -> bool:
    return int(state["buffer_index"]) == 0


def _clamp_cursor(state: dict[str, Any]) -> None:
    lines = _editor_lines(state)
    row = max(0, min(int(state["row"]), max(0, len(lines) - 1)))
    state["row"] = row
    state["col"] = max(0, min(int(state["col"]), len(lines[row]) if lines else 0))


def _push_undo(state: dict[str, Any]) -> None:
    state["undo"].append({"buffer": list(state["buffer"]), "row": state["row"], "col": state["col"]})
    state["undo"] = state["undo"][-30:]


def _switch_buffer(state: dict[str, Any], delta: int) -> None:
    count = len(state["reference_buffers"]) + 1
    state["buffer_index"] = (int(state["buffer_index"]) + delta) % count
    state["visited_buffers"].add(state["buffer_index"])
    state["buffer_switches"] += 1
    state["row"] = 0
    state["col"] = 0
    state["pending"] = ""


def _advance_layer(state: dict[str, Any]) -> None:
    current = _layer_name(state)
    if current in {"editor", "complete"}:
        return
    state["exit_log"].append(current)
    state["layer_index"] += 1
    state["ssh_input"] = ""
    state["mux_pending"] = False


def _apply_editor_key(state: dict[str, Any], key: str, ctrl: bool, alt: bool, meta: bool) -> None:
    mode = state["mode"]
    if mode == "insert":
        if not _editor_writable(state):
            raise ValueError("insert mode entered on read-only reference")
        if key == "Escape":
            state["mode"] = "normal"
            state["pending"] = ""
        elif key == "Backspace":
            if state["col"] > 0:
                _push_undo(state)
                row, col = state["row"], state["col"]
                line = state["buffer"][row]
                state["buffer"][row] = line[: col - 1] + line[col:]
                state["col"] -= 1
        elif key == "Delete":
            row, col = state["row"], state["col"]
            line = state["buffer"][row]
            if col < len(line):
                _push_undo(state)
                state["buffer"][row] = line[:col] + line[col + 1 :]
        elif len(key) == 1 and not ctrl and not alt and not meta:
            _push_undo(state)
            row, col = state["row"], state["col"]
            line = state["buffer"][row]
            state["buffer"][row] = line[:col] + key + line[col:]
            state["col"] += 1
            state["inserted_chars"] += 1
        _clamp_cursor(state)
        return

    if mode == "command":
        if key == "Escape":
            state["mode"] = "normal"
            state["command"] = ""
        elif key == "Backspace":
            state["command"] = state["command"][:-1]
        elif key == "Enter":
            command = state["command"]
            state["command_history"].append(command)
            state["command"] = ""
            state["mode"] = "normal"
            if command in {"bn", "bnext"}:
                _switch_buffer(state, 1)
            elif command in {"bp", "bprevious"}:
                _switch_buffer(state, -1)
            elif (
                command == "wq"
                and _editor_writable(state)
                and state["visited_buffers"] == set(range(len(state["reference_buffers"]) + 1))
                and state["buffer"] == state["target_buffer"]
            ):
                state["saved"] = True
                state["layer_index"] = 0
            else:
                state["command_errors"] += 1
        elif len(key) == 1 and not ctrl and not alt and not meta:
            state["command"] += key
        return

    if key == "Escape":
        state["pending"] = ""
        return
    if ctrl or alt or meta:
        state["pending"] = ""
        return
    if key == ":":
        state["mode"] = "command"
        state["command"] = ""
        state["pending"] = ""
        return
    if key == "g":
        if state["pending"] == "g":
            state["row"] = 0
            state["col"] = 0
            state["pending"] = ""
        else:
            state["pending"] = "g"
        return
    if not _editor_writable(state) and key in {"c", "i", "a", "x", "u"}:
        state["pending"] = ""
        return
    if key == "c":
        if state["pending"] == "c":
            _push_undo(state)
            state["buffer"][state["row"]] = ""
            state["col"] = 0
            state["mode"] = "insert"
            state["pending"] = ""
            state["clear_count"] += 1
        else:
            state["pending"] = "c"
        return
    state["pending"] = ""
    if key in {"j", "ArrowDown"}:
        state["row"] += 1
    elif key in {"k", "ArrowUp"}:
        state["row"] -= 1
    elif key in {"h", "ArrowLeft"}:
        state["col"] -= 1
    elif key in {"l", "ArrowRight"}:
        state["col"] += 1
    elif key in {"0", "Home"}:
        state["col"] = 0
    elif key in {"$", "End"}:
        state["col"] = len(_editor_lines(state)[state["row"]])
    elif key == "i":
        state["mode"] = "insert"
    elif key == "a":
        state["col"] = min(len(state["buffer"][state["row"]]), state["col"] + 1)
        state["mode"] = "insert"
    elif key == "x":
        row, col = state["row"], state["col"]
        line = state["buffer"][row]
        if col < len(line):
            _push_undo(state)
            state["buffer"][row] = line[:col] + line[col + 1 :]
    elif key == "u" and state["undo"]:
        snapshot = state["undo"].pop()
        state["buffer"] = list(snapshot["buffer"])
        state["row"] = snapshot["row"]
        state["col"] = snapshot["col"]
    _clamp_cursor(state)


def _apply_event(state: dict[str, Any], event: dict[str, Any]) -> None:
    key = str(event.get("key") or "")
    ctrl = event.get("ctrl") is True
    alt = event.get("alt") is True
    meta = event.get("meta") is True
    current = _layer_name(state)
    if current == "editor":
        _apply_editor_key(state, key, ctrl, alt, meta)
    elif current == "pager":
        if key.lower() == "q" and not ctrl and not alt and not meta:
            _advance_layer(state)
    elif current == "job":
        if key.lower() == "c" and ctrl and not alt and not meta:
            _advance_layer(state)
    elif current == "ssh":
        if key == "Backspace":
            state["ssh_input"] = state["ssh_input"][:-1]
        elif key == "Enter":
            if state["ssh_input"].strip() == "exit":
                _advance_layer(state)
            else:
                state["command_errors"] += 1
                state["ssh_input"] = ""
        elif len(key) == 1 and not ctrl and not alt and not meta:
            state["ssh_input"] += key
    elif current == "mux":
        if key.lower() == "b" and ctrl and not alt and not meta:
            state["mux_pending"] = True
        elif state["mux_pending"] and key.lower() == "d" and not ctrl and not alt and not meta:
            _advance_layer(state)
        else:
            state["mux_pending"] = False


def _initial_state(ground_truth: dict[str, Any]) -> dict[str, Any]:
    initial = ground_truth.get("initial_buffer")
    target = ground_truth.get("target_buffer")
    references = ground_truth.get("reference_buffers")
    layers = ground_truth.get("layer_order")
    if not isinstance(initial, list) or not isinstance(target, list) or not isinstance(references, list) or not isinstance(layers, list):
        raise ValueError("terminal contract is malformed")
    if len(initial) != len(target) or len(initial) != 6 or len(references) != 3:
        raise ValueError("multi-buffer terminal shape is invalid")
    if not all(isinstance(item, str) for item in [*initial, *target, *layers]):
        raise ValueError("terminal contract contains non-text values")
    if any(not isinstance(item, dict) or not isinstance(item.get("name"), str) or not isinstance(item.get("lines"), list) or not all(isinstance(line, str) for line in item["lines"]) for item in references):
        raise ValueError("reference-buffer contract is malformed")
    return {
        "buffer": list(initial),
        "target_buffer": list(target),
        "reference_buffers": deepcopy(references),
        "buffer_index": 0,
        "visited_buffers": {0},
        "buffer_switches": 0,
        "layer_order": list(layers),
        "layer_index": -1,
        "mode": "normal",
        "row": 0,
        "col": 0,
        "pending": "",
        "command": "",
        "undo": [],
        "saved": False,
        "exit_log": [],
        "ssh_input": "",
        "mux_pending": False,
        "clear_count": 0,
        "inserted_chars": 0,
        "command_errors": 0,
        "command_history": [],
    }


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    challenge_id = str(ground_truth.get("challenge_id") or "")
    task_id = str(ground_truth.get("task_id") or "")
    if str(payload.get("mechanic_id") or "") != MECHANIC_ID or str(ground_truth.get("mechanic_id") or "") != MECHANIC_ID:
        return {"graded": True, "passed": False, "feedback": "mechanic mismatch"}
    if not challenge_id or str(payload.get("challenge_id") or "") != challenge_id or str(public_state.get("challenge_id") or "") != challenge_id:
        return {"graded": True, "passed": False, "feedback": "stale challenge"}
    if not task_id or str(payload.get("task_id") or "") != task_id or str(public_state.get("task_id") or "") != task_id:
        return {"graded": True, "passed": False, "feedback": "task identity mismatch"}
    for field in ("initial_buffer", "target_buffer", "reference_buffers", "layer_order", "host"):
        if public_state.get(field) != ground_truth.get(field):
            return {"graded": True, "passed": False, "feedback": f"public/private terminal {field} contract skew"}
    try:
        state = _initial_state(ground_truth)
    except ValueError as exc:
        return {"graded": True, "passed": False, "feedback": str(exc)}

    events = payload.get("events")
    if not isinstance(events, list) or not (1 <= len(events) <= 2000):
        return {"graded": True, "passed": False, "feedback": "keystroke transcript is missing or too long"}
    allowed_fields = {"sequence", "key", "ctrl", "shift", "alt", "meta", "layer_before", "mode_before", "layer_after", "mode_after"}
    for sequence, event in enumerate(events, start=1):
        if not isinstance(event, dict) or event.get("sequence") != sequence or set(event) != allowed_fields:
            return {"graded": True, "passed": False, "feedback": f"key event {sequence} has invalid sequence or schema"}
        key = event.get("key")
        if not isinstance(key, str) or not key or len(key) > 24 or any(not isinstance(event.get(flag), bool) for flag in ("ctrl", "shift", "alt", "meta")):
            return {"graded": True, "passed": False, "feedback": f"key event {sequence} has invalid key data"}
        if event.get("layer_before") != _layer_name(state) or event.get("mode_before") != _mode_name(state):
            return {"graded": True, "passed": False, "feedback": f"key event {sequence} disagrees before replay"}
        try:
            _apply_event(state, event)
        except ValueError as exc:
            return {"graded": True, "passed": False, "feedback": f"key event {sequence}: {exc}"}
        if event.get("layer_after") != _layer_name(state) or event.get("mode_after") != _mode_name(state):
            return {"graded": True, "passed": False, "feedback": f"key event {sequence} disagrees after replay"}

    final_state = {
        "buffer": list(state["buffer"]),
        "saved": state["saved"],
        "layer_index": state["layer_index"],
        "exit_log": list(state["exit_log"]),
        "mode": _mode_name(state),
        "buffer_index": state["buffer_index"],
        "visited_buffers": sorted(state["visited_buffers"]),
        "buffer_switches": state["buffer_switches"],
    }
    if payload.get("final_state") != final_state:
        return {"graded": True, "passed": False, "feedback": "claimed terminal state does not match key replay"}
    layers = state["layer_order"]
    all_buffers = set(range(len(state["reference_buffers"]) + 1))
    passed = (
        state["saved"] is True
        and state["buffer"] == state["target_buffer"]
        and state["visited_buffers"] == all_buffers
        and state["buffer_switches"] >= len(state["reference_buffers"]) + 1
        and state["layer_index"] == len(layers)
        and state["exit_log"] == layers
        and state["clear_count"] >= len(state["target_buffer"])
        and state["inserted_chars"] >= sum(len(line) for line in state["target_buffer"])
        and "wq" in state["command_history"]
    )
    return {
        "graded": True,
        "passed": passed,
        "score": 100 if passed else 0,
        "feedback": (
            f"buffers visited {len(state['visited_buffers'])}/{len(all_buffers)}; "
            f"fields rewritten {state['clear_count']}/{len(state['target_buffer'])}; "
            f"layers escaped {len(state['exit_log'])}/{len(layers)}; key events {len(events)}"
        ),
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    del public_state
    return {
        "target_buffer": deepcopy(ground_truth.get("target_buffer") or []),
        "reference_buffers": deepcopy(ground_truth.get("reference_buffers") or []),
        "layer_order": list(ground_truth.get("layer_order") or []),
        "instructions": ["visit :bn buffers", "return with :bn/:bp", "repair manifest with cc", ":wq", "escape each visible outer layer"],
        "answers": [],
    }
